File: src/lib/focus-score.ts
// Focus Score engine — pure functions, no React or DB deps.
// Spec: S_focus = max(0, min(100, (T_focus/T_target)*100 - sum(P_breach)))
// XP   = floor(S_focus * (T_focus/60) * M_tier)
//
// Brief §3: durations flow through as fractional seconds (ms-precision) and
// are only floored at the *boundary* where we hand integers to Postgres.

/**
 * Stamped on every completed session so a historical score can be interpreted
 * with the rules that actually produced it.
 *
 * Without this, changing the formula silently reinterprets every past result:
 * a leaderboard would compare v1 and v2 scores as though they meant the same
 * thing, and nobody could tell which. Bump this whenever the tier boundaries,
 * penalties or XP multipliers change.
 *
 * v2 — tier and XP derive from the unrounded score. In v1 the rounded value
 * fed the tier lookup, so a raw 84.5 rounded to 85 and doubled the XP
 * multiplier at a boundary.
 */
export const SCORING_VERSION = 2;

export type Tier =
  | "flow" // 95-100
  | "pristine" // 85-94
  | "steady" // 70-84
  | "fragmented" // 40-69
  | "compromised"; // 0-39

export interface TierMeta {
  key: Tier;
  label: string;
  hex: string;
  multiplier: number;
}

export const TIERS: Record<Tier, TierMeta> = {
  flow: { key: "flow", label: "Flow State", hex: "#06B6D4", multiplier: 1.5 },
  pristine: { key: "pristine", label: "Pristine Focus", hex: "#10B981", multiplier: 1.0 },
  steady: { key: "steady", label: "Steady Ambient", hex: "#F59E0B", multiplier: 0.5 },
  fragmented: { key: "fragmented", label: "Fragmented Attention", hex: "#F97316", multiplier: 0.0 },
  compromised: {
    key: "compromised",
    label: "Protocol Compromised",
    hex: "#EF4444",
    multiplier: 0.0,
  },
};

export const ABANDONMENT_GRACE_SECONDS = 15;
export const MINOR_PENALTY = 10;
export const SEVERE_PENALTY = 40;

export function tierForScore(score: number): TierMeta {
  if (score >= 95) return TIERS.flow;
  if (score >= 85) return TIERS.pristine;
  if (score >= 70) return TIERS.steady;
  if (score >= 40) return TIERS.fragmented;
  return TIERS.compromised;
}

export interface BreachRecord {
  severity: "minor" | "severe";
}

export interface ScoreInput {
  /** target in seconds (may be fractional) */
  targetSeconds: number;
  /** focused time in seconds (may be fractional / ms-precise) */
  focusSeconds: number;
  breaches: BreachRecord[];
  /** continuous time past grace window after a severe breach (fractional ok) */
  abandonmentSeconds?: number;
}

export interface ScoreResult {
  score: number;
  xp: number;
  tier: TierMeta;
  penalty: number;
  abandonmentPenalty: number;
  /** echoed back at integer second-precision for DB persistence */
  focusSecondsInt: number;
  /** Ruleset that produced this result — persist it alongside the score. */
  scoringVersion: number;
}

export function computeFocusScore({
  targetSeconds,
  focusSeconds,
  breaches,
  abandonmentSeconds = 0,
}: ScoreInput): ScoreResult {
  const target = Math.max(1, targetSeconds);
  const focus = Math.max(0, Math.min(focusSeconds, target));

  const breachPenalty = breaches.reduce(
    (sum, b) => sum + (b.severity === "severe" ? SEVERE_PENALTY : MINOR_PENALTY),
    0,
  );
  const abandonmentPenalty = Math.max(
    0,
    Math.floor(abandonmentSeconds - ABANDONMENT_GRACE_SECONDS),
  );
  const penalty = breachPenalty + abandonmentPenalty;

  // Keep ms-precision through the multiplications — only round at the end.
  const raw = (focus / target) * 100 - penalty;

  // Rounding is a *display* concern and must not decide rewards. Previously
  // the rounded value fed both the tier lookup and the XP formula, so a raw
  // 84.5 rounded up to 85, crossed the "pristine" boundary, and doubled the
  // multiplier — a half-point of real performance became a 2x reward jump.
  // Tier and XP are now derived from the unrounded score; only `score` is
  // rounded, and only because it is the number shown to the user.
  const rawClamped = Math.max(0, Math.min(100, raw));
  const score = Math.round(rawClamped);
  const tier = tierForScore(rawClamped);
  const xp = Math.floor(rawClamped * (focus / 60) * tier.multiplier);

  return {
    score,
    xp,
    tier,
    penalty: breachPenalty,
    abandonmentPenalty,
    focusSecondsInt: Math.floor(focus),
    scoringVersion: SCORING_VERSION,
  };
}

File: android/app/src/main/java/app/stackd/feature/room/session/FocusScore.kt
package app.stackd.feature.room.session

import kotlin.math.floor
import kotlin.math.roundToInt

/**
 * Ported 1:1 from the web's `src/lib/focus-score.ts`.
 *
 * Pure Kotlin — no Android types — so the reward maths is unit-testable on the
 * JVM, the same split [BreachRules] uses. A score computed here and a score
 * computed by the web app for the same session must agree exactly; they land in
 * the same `focus_history` table and the same leaderboards.
 */
object FocusScore {

    /**
     * Stamped on every completed session so a historical score stays
     * interpretable under the rules that produced it. Bump whenever tier
     * boundaries, penalties, or XP multipliers change.
     *
     * v2 — tier and XP derive from the *unrounded* score. In v1 the rounded
     * value fed the tier lookup, so a raw 84.5 rounded to 85 and doubled the
     * XP multiplier at a boundary.
     */
    const val SCORING_VERSION = 2

    const val ABANDONMENT_GRACE_SECONDS = 15
    const val MINOR_PENALTY = 10
    const val SEVERE_PENALTY = 40

    enum class Tier(
        val key: String,
        val label: String,
        val hex: Long,
        val multiplier: Double,
    ) {
        FLOW("flow", "Flow State", 0xFF06B6D4, 1.5),
        PRISTINE("pristine", "Pristine Focus", 0xFF10B981, 1.0),
        STEADY("steady", "Steady Ambient", 0xFFF59E0B, 0.5),
        FRAGMENTED("fragmented", "Fragmented Attention", 0xFFF97316, 0.0),
        COMPROMISED("compromised", "Protocol Compromised", 0xFFEF4444, 0.0),
    }

    fun tierForScore(score: Double): Tier = when {
        score >= 95 -> Tier.FLOW
        score >= 85 -> Tier.PRISTINE
        score >= 70 -> Tier.STEADY
        score >= 40 -> Tier.FRAGMENTED
        else -> Tier.COMPROMISED
    }

    data class Result(
        val score: Int,
        val xp: Int,
        val tier: Tier,
        val penalty: Int,
        val abandonmentPenalty: Int,
        /** Echoed back at integer-second precision, for the DB boundary. */
        val focusSecondsInt: Int,
        val scoringVersion: Int = SCORING_VERSION,
    )

    /**
     * `S_focus = clamp(0, 100, (T_focus / T_target) * 100 - Σ penalties)`,
     * `XP = floor(S_focus * (T_focus / 60) * tierMultiplier)`.
     *
     * Durations arrive as fractional seconds (ms precision preserved from the
     * timestamps) and are only floored where integers are handed to Postgres.
     */
    fun compute(
        targetSeconds: Double,
        focusSeconds: Double,
        severeBreaches: Int,
        minorBreaches: Int,
        abandonmentSeconds: Double = 0.0,
    ): Result {
        val target = targetSeconds.coerceAtLeast(1.0)
        val focus = focusSeconds.coerceIn(0.0, target)

        val breachPenalty = severeBreaches * SEVERE_PENALTY + minorBreaches * MINOR_PENALTY
        val abandonmentPenalty =
            floor(abandonmentSeconds - ABANDONMENT_GRACE_SECONDS).coerceAtLeast(0.0).toInt()
        val penalty = breachPenalty + abandonmentPenalty

        val raw = (focus / target) * 100 - penalty
        val rawClamped = raw.coerceIn(0.0, 100.0)

        // Rounding is a display concern and must not decide rewards — tier and
        // XP read the unrounded value, only `score` is rounded.
        val tier = tierForScore(rawClamped)
        return Result(
            score = rawClamped.roundToInt(),
            xp = floor(rawClamped * (focus / 60.0) * tier.multiplier).toInt(),
            tier = tier,
            penalty = breachPenalty,
            abandonmentPenalty = abandonmentPenalty,
            focusSecondsInt = floor(focus).toInt(),
        )
    }
}
